pkg/metrics/instance/group_manager.go

```diff
@@ -30,11 +37,152 @@ type GroupManager struct {
+	groupsInConfigs := make(map[string]groupedConfigs)
+
+	// Combined group of current and new configs
+	combinedConfigs := m.getCombinedConfigs(configs)
+
+	oldConfigs := make([]Config, 0)
+	// In case of an error in applying the new configs, we need to grab the old config for rollback
+	copy(oldConfigs, m.activeConfigs)
+	activeConfigs := make([]Config, 0)
+	groupLookup := make(map[string]string)
+	// Iterate through the combined configurations
+	for _, c := range combinedConfigs {
+		groupName, err := hashConfig(c)
+		if err != nil {
+			failed = append(failed, BatchFailure{
+				Err:    err,
+				Config: c,
+			})
+			level.Error(m.log).Log("err", fmt.Sprintf("failed to get group name for config %s: %s", c.Name, err))
+			continue
+		}
+		if _, exists := groupsInConfigs[groupName]; !exists {
+			groupsInConfigs[groupName] = make(groupedConfigs)
+		}
+		groupsInConfigs[groupName][c.Name] = c
+		groupLookup[c.Name] = groupName
+		activeConfigs = append(activeConfigs, c)
+	}
+	// If we have group that no longer exist then we need to delete them
+	for groupName := range m.groups {
+		if _, exists := groupsInConfigs[groupName]; !exists {
+			if err := m.inner.DeleteConfig(groupName); err != nil {
+				level.Error(m.log).Log("err", fmt.Sprintf("failed to delete group named  %s with error %s", groupName, err))
+			}
+		}
+	}
+	groups := make(map[string]groupedConfigs)
+	mergedHashes := make(map[string]int)
+	// Now that we have grouped all the new and existing configurations we can apply them
+	for groupName, grouped := range groupsInConfigs {
+		mergedConfig, err := groupConfigs(groupName, grouped)
+		if err != nil {
+			level.Error(m.log).Log("err", fmt.Sprintf("failed to group configs with groupname  %s with error %s", groupName, err))
+		}
+		// If we have the exact same config no need to reload it
+		newHash, err := createMergedConfigHash(mergedConfig)
+		if err != nil {
+			level.Error(m.log).Log("err", fmt.Sprintf("failed to create merged hash named  %s with error %s", groupName, err))
+			continue
+		}
+		groups[groupName] = grouped
+		mergedHashes[newHash] = 0
+		if _, exists := m.mergedConfigHashes[newHash]; exists {
+			continue
+		}
+
+		// Something terrible happened so roll back to the old configurations, this ensures that at least the agent
+		// always has a valid configuration. If we are already in a rollback then dont try again
+		if err != nil && !isRollback {
+			m.rollbackConfigs(oldConfigs)
+			return CreateBatchApplyErrorOrNil(failed, err)
+		}
+		err = m.inner.ApplyConfig(mergedConfig)
+		if err != nil && !isRollback {
+			m.rollbackConfigs(oldConfigs)
+			return CreateBatchApplyErrorOrNil(failed, err)
+		}
+	}
+	m.groups = groups
+	m.activeConfigs = activeConfigs
+	m.groupLookup = groupLookup
+	m.mergedConfigHashes = mergedHashes
+
+	return CreateBatchApplyErrorOrNil(failed, nil)
+}
+
+func createMergedConfigHash(mergedConfig Config) (string, error) {
+	bb, err := yaml.Marshal(mergedConfig)
```

Comment:
nit: `yaml.Marshal` for configs has overhead because it creates a yaml.Map. Use `MarshalConfig` directly to reduce allocations.